File: android/measure/src/main/java/sh/measure/android/cel/CelTokenizer.kt
package sh.measure.android.cel

/**
 * Lexical analyzer for CEL expressions that converts raw text into a stream of meaningful tokens.
 * Handles operators, literals, identifiers, and punctuation while preserving source positions
 * for accurate error reporting.
 */
class CelTokenizer {

    /**
     * Token representation containing type information, textual value, and source position.
     * Position tracking enables precise error reporting during parsing and evaluation.
     */
    data class Token(
        val type: TokenType,
        val value: String,
        val position: Int,
    )

    /**
     * Enumeration of all supported token types in CEL expressions.
     * Organized by category: literals, operators, punctuation, and special markers.
     */
    enum class TokenType {
        // Literals
        IDENTIFIER,      // Field names, function names, variables
        STRING,          // String literals in double quotes
        NUMBER,          // Numeric literals (integers and decimals)
        BOOLEAN,         // Boolean literals 'true' and 'false'
        NULL,            // Null literal

        // Comparison operators
        EQUALS,          // ==
        NOT_EQUALS,      // !=
        GREATER_THAN,    // >
        LESS_THAN,       // <
        GREATER_EQUAL,   // >=
        LESS_EQUAL,      // <=

        // String operations
        CONTAINS,        // contains
        STARTS_WITH,     // startsWith
        ENDS_WITH,       // endsWith

        // Logical operators
        AND,             // &&
        OR,              // ||

        // Punctuation
        DOT,             // .
        LPAREN,          // (
        RPAREN,          // )
        COMMA,           // ,

        // Special
        EOF              // End of file marker
    }

    /**
     * Tokenizes a CEL expression string into a list of tokens.
     *
     * Performs lexical analysis by scanning the input character by character,
     * recognizing patterns for different token types. Multi-character operators
     * are detected before single-character ones to avoid ambiguous tokenization.
     *
     * @param expression The CEL expression string to tokenize
     * @return List of tokens representing the lexical structure
     * @throws IllegalArgumentException if an unexpected character is encountered
     */
    fun tokenize(expression: String): List<Token> {
        val tokens = mutableListOf<Token>()
        var i = 0

        while (i < expression.length) {
            when {
                expression[i].isWhitespace() -> {
                    i++
                }

                // Single character tokens
                expression[i] == '(' -> {
                    tokens.add(Token(TokenType.LPAREN, "(", i))
                    i++
                }

                expression[i] == ')' -> {
                    tokens.add(Token(TokenType.RPAREN, ")", i))
                    i++
                }

                expression[i] == '.' -> {
                    tokens.add(Token(TokenType.DOT, ".", i))
                    i++
                }

                expression[i] == ',' -> {
                    tokens.add(Token(TokenType.COMMA, ",", i))
                    i++
                }

                // String literals
                expression[i] == '"' -> {
                    val start = i
                    i++ // skip opening quote
                    while (i < expression.length && expression[i] != '"') {
                        i++
                    }
                    if (i >= expression.length) {
                        throw IllegalArgumentException("Unterminated string literal starting at position $start")
                    }
                    i++ // skip closing quote
                    tokens.add(
                        Token(
                            TokenType.STRING,
                            expression.substring(start + 1, i - 1),
                            start
                        )
                    )
                }

                // Numbers
                expression[i].isDigit() -> {
                    val start = i
                    while (i < expression.length && (expression[i].isDigit() || expression[i] == '.')) {
                        i++
                    }
                    tokens.add(Token(TokenType.NUMBER, expression.substring(start, i), start))
                }

                // Identifiers and keywords
                expression[i].isLetter() || expression[i] == '_' -> {
                    val start = i
                    while (i < expression.length && (expression[i].isLetterOrDigit() || expression[i] == '_')) {
                        i++
                    }
                    val value = expression.substring(start, i)
                    val type = determineKeywordType(value)
                    tokens.add(Token(type, value, start))
                }

                // Multi-character operators (must be checked before single-character ones)
                expression.startsWith("==", i) -> {
                    tokens.add(Token(TokenType.EQUALS, "==", i))
                    i += 2
                }

                expression.startsWith("!=", i) -> {
                    tokens.add(Token(TokenType.NOT_EQUALS, "!=", i))
                    i += 2
                }

                expression.startsWith(">=", i) -> {
                    tokens.add(Token(TokenType.GREATER_EQUAL, ">=", i))
                    i += 2
                }

                expression.startsWith("<=", i) -> {
                    tokens.add(Token(TokenType.LESS_EQUAL, "<=", i))
                    i += 2
                }

                expression.startsWith("&&", i) -> {
                    tokens.add(Token(TokenType.AND, "&&", i))
                    i += 2
                }

                expression.startsWith("||", i) -> {
                    tokens.add(Token(TokenType.OR, "||", i))
                    i += 2
                }

                // Single-character comparison operators
                expression[i] == '>' -> {
                    tokens.add(Token(TokenType.GREATER_THAN, ">", i))
                    i++
                }

                expression[i] == '<' -> {
                    tokens.add(Token(TokenType.LESS_THAN, "<", i))
                    i++
                }

                else -> {
                    throw IllegalArgumentException("Unexpected character: '${expression[i]}' at position $i")
                }
            }
        }

        tokens.add(Token(TokenType.EOF, "", expression.length))
        return tokens
    }

    /**
     * Determines the appropriate token type for identifier-like strings.
     * Recognizes reserved keywords and special identifiers, classifying them
     * into their appropriate token types rather than generic identifiers.
     */
    private fun determineKeywordType(value: String): TokenType = when (value) {
        "true", "false" -> TokenType.BOOLEAN
        "null", "NULL" -> TokenType.NULL
        "contains" -> TokenType.CONTAINS
        "startsWith" -> TokenType.STARTS_WITH
        "endsWith" -> TokenType.ENDS_WITH
        else -> TokenType.IDENTIFIER
    }
}

File: frontend/dashboard/app/utils/cel_tokenizer.ts
// Enhanced error types for better debugging
export class CelParseError extends Error {
  constructor(
    message: string,
    public position: number,
    public token?: string,
    public expected?: string[]
  ) {
    super(`${message} at position ${position}${token ? ` (found: '${token}')` : ''}${expected ? ` (expected: ${expected.join(' or ')})` : ''}`)
    this.name = 'CelParseError'
  }
}

// Token types
export enum TokenType {
  // Literals
  IDENTIFIER = 'IDENTIFIER',
  STRING = 'STRING',
  NUMBER = 'NUMBER',
  BOOLEAN = 'BOOLEAN',
  NULL = 'NULL',

  // Operators
  EQUALS = 'EQUALS',
  NOT_EQUALS = 'NOT_EQUALS',
  GREATER_THAN = 'GREATER_THAN',
  LESS_THAN = 'LESS_THAN',
  GREATER_EQUAL = 'GREATER_EQUAL',
  LESS_EQUAL = 'LESS_EQUAL',

  // String operations
  CONTAINS = 'CONTAINS',
  STARTS_WITH = 'STARTS_WITH',

  // Logical
  AND = 'AND',
  OR = 'OR',

  // Punctuation
  DOT = 'DOT',
  LPAREN = 'LPAREN',
  RPAREN = 'RPAREN',
  COMMA = 'COMMA',

  // Special
  EOF = 'EOF'
}

export interface Token {
  type: TokenType
  value: string
  position: number
}

// Tokenizer
export class CelTokenizer {
  private input: string
  private position = 0
  private tokens: Token[] = []

  constructor(input: string) {
    this.input = input.trim()
  }

  tokenize(): Token[] {
    this.position = 0
    this.tokens = []

    while (this.position < this.input.length) {
      this.skipWhitespace()

      if (this.position >= this.input.length) break

      const char = this.input[this.position]
      const startPos = this.position

      // Multi-character operators
      if (this.position + 1 < this.input.length) {
        const twoChar = this.input.slice(this.position, this.position + 2);
        const tokenType = this.getTwoCharTokenType(twoChar)
        if (tokenType) {
          this.tokens.push({ type: tokenType, value: twoChar, position: startPos })
          this.position += 2
          continue
        }
      }

      // Single character tokens
      const singleTokenType = this.getSingleCharTokenType(char)
      if (singleTokenType) {
        this.tokens.push({ type: singleTokenType, value: char, position: startPos })
        this.position++
        continue
      }

      // String literals
      if (char === '"') {
        const stringValue = this.parseStringLiteral()
        this.tokens.push({ type: TokenType.STRING, value: stringValue, position: startPos })
        continue
      }

      // Numbers
      if (/\d/.test(char)) {
        const numberValue = this.parseNumber()
        this.tokens.push({ type: TokenType.NUMBER, value: numberValue, position: startPos })
        continue
      }

      // Identifiers and keywords
      if (/[a-zA-Z_]/.test(char)) {
        const identifier = this.parseIdentifier()
        const tokenType = this.getKeywordTokenType(identifier)
        this.tokens.push({ type: tokenType, value: identifier, position: startPos })
        continue
      }

      throw new CelParseError(`Unexpected character '${char}'`, this.position)
    }

    this.tokens.push({ type: TokenType.EOF, value: '', position: this.position })
    return this.tokens
  }

  private skipWhitespace(): void {
    while (this.position < this.input.length && /\s/.test(this.input[this.position])) {
      this.position++
    }
  }

  private getTwoCharTokenType(twoChar: string): TokenType | null {
    switch (twoChar) {
      case '==': return TokenType.EQUALS
      case '!=': return TokenType.NOT_EQUALS
      case '>=': return TokenType.GREATER_EQUAL
      case '<=': return TokenType.LESS_EQUAL
      case '&&': return TokenType.AND
      case '||': return TokenType.OR
      default: return null
    }
  }

  private getSingleCharTokenType(char: string): TokenType | null {
    switch (char) {
      case '(': return TokenType.LPAREN
      case ')': return TokenType.RPAREN
      case '.': return TokenType.DOT
      case ',': return TokenType.COMMA
      case '>': return TokenType.GREATER_THAN
      case '<': return TokenType.LESS_THAN
      default: return null
    }
  }

  private getKeywordTokenType(identifier: string): TokenType {
    switch (identifier) {
      case 'true':
      case 'false':
        return TokenType.BOOLEAN
      case 'null':
        return TokenType.NULL
      case 'contains':
        return TokenType.CONTAINS
      case 'startsWith':
        return TokenType.STARTS_WITH
      default:
        return TokenType.IDENTIFIER
    }
  }

  private parseStringLiteral(): string {
    this.position++ // Skip opening quote
    let value = ''

    while (this.position < this.input.length && this.input[this.position] !== '"') {
      if (this.input[this.position] === '\\' && this.position + 1 < this.input.length) {
        // Handle escape sequences
        this.position++
        switch (this.input[this.position]) {
          case 'n': value += '\n'; break
          case 't': value += '\t'; break
          case 'r': value += '\r'; break
          case '\\': value += '\\'; break
          case '"': value += '"'; break
          default: value += this.input[this.position]; break
        }
      } else {
        value += this.input[this.position]
      }
      this.position++
    }

    if (this.position >= this.input.length) {
      throw new CelParseError('Unterminated string literal', this.position - value.length - 1)
    }

    this.position++
    return value
  }

  private parseNumber(): string {
    let value = ''
    let hasDecimal = false

    while (this.position < this.input.length) {
      const char = this.input[this.position]
      if (/\d/.test(char)) {
        value += char
        this.position++
      } else if (char === '.' && !hasDecimal) {
        hasDecimal = true
        value += char
        this.position++
      } else {
        break
      }
    }

    return value
  }

  private parseIdentifier(): string {
    let value = ''

    while (this.position < this.input.length && /[a-zA-Z0-9_]/.test(this.input[this.position])) {
      value += this.input[this.position]
      this.position++
    }

    return value
  }
}